File: Lab_7/main2.py
import math


class PersistentSegmentTree:
    def __init__(self, ar):
        self._n = len(ar)

        LOG_N = int(math.ceil(math.log(self._n, 2))) + 1

        self._node_size = 3
        self._layer_size = self._node_size * 2 * self._n
        self._tree = [None] * (LOG_N * self._layer_size)
        self._last_node_idx = [i * self._layer_size for i in range(LOG_N)]

        self._build_tree(0, ar, 0, self._n - 1)

    def _build_tree(self, v, ar, tl, tr):
        if tl == tr:
            cur_layer = v // self._layer_size
            self._last_node_idx[cur_layer] += self._node_size

            self._tree[v] = ar[tl]
            return self._tree[v]
        else:
            tm = (tl + tr) // 2

            cur_layer = v // self._layer_size
            left_idx = self._last_node_idx[cur_layer + 1]
            right_idx = self._last_node_idx[cur_layer + 1] + self._node_size

            left = self._build_tree(left_idx, ar, tl, tm)
            right = self._build_tree(right_idx, ar, tm + 1, tr)

            self._last_node_idx[cur_layer] += self._node_size

            self._tree[v] = left + right
            self._tree[v + 1] = left_idx
            self._tree[v + 2] = right_idx
            return self._tree[v]

    def _set(self, v, tl, tr, pos, val):
        if tl == tr:
            cur_layer = v // self._layer_size
            new_v = self._last_node_idx[cur_layer]
            self._tree[new_v] = val
            self._last_node_idx[cur_layer] += self._node_size
            return new_v, self._tree[new_v]
        else:
            tm = (tr + tl) // 2

            cur_layer = v // self._layer_size

            if pos <= tm:
                left_idx, left_val = self._set(self._tree[v + 1], tl, tm, pos, val)

                right_idx = self._tree[v + 2]
                right_val = self._tree[right_idx]
            else:
                left_idx = self._tree[v + 1]
                left_val = self._tree[left_idx]

                right_idx, right_val = self._set(self._tree[v + 2], tm + 1, tr, pos, val)

            new_v = self._last_node_idx[cur_layer]
            self._tree[new_v] = left_val + right_val
            self._tree[new_v + 1] = left_idx
            self._tree[new_v + 2] = right_idx
            self._last_node_idx[cur_layer] += self._node_size
            return new_v, self._tree[new_v]

    def set(self, pos, val):
        self._set(self._last_node_idx[0] - self._node_size, 0, self._n - 1, pos, val)

    def _get_sum(self, v, tl, tr, l, r):
        if l > r:
            return 0
        if (l == tl) and (r == tr):
            return self._tree[v]

        tm = (tr + tl) // 2
        return self._get_sum(self._tree[v + 1], tl, tm, l, min(r, tm)) + \
               self._get_sum(self._tree[v + 2], tm + 1, tr, max(l, tm + 1), r)

    def get_sum(self, root_idx, left, right):
        return self._get_sum(root_idx * self._node_size, 0, self._n - 1, left, right)


def print_tree(tree):
    print("\nTree:")
    for idx, val in enumerate(tree._tree):
        print(val, end=' ')
        if idx % tree._layer_size == tree._layer_size - 1:
            print()
    print("End Tree\n")


def binary_search(ar, x):
    l, r = 0, len(ar)

    while r - 1 != l:
        tmp = (r + l) // 2
        if ar[tmp][0] < x:
            l = tmp
        else:
            r = tmp

    if ar[l][0] >= x:
        return -1

    return l


if __name__ == '__main__':
    N, Q = list(map(int, input().split()))
    A = list(map(int, input().split()))

    tree = PersistentSegmentTree([0] * len(A))

    ar = sorted(zip(A, range(N)), key=lambda x: x[0])

    # print_tree(tree)

    for x in ar:
        tree.set(x[1], 1)
        # print_tree(tree)

    # print(ar)

    # for i in range(6):
    #     print(tree.get_sum(i, 0, len(A) - 1))

    for _ in range(Q):
        l, r, x, y = list(map(int, input().split()))

        idx_x = binary_search(ar, x)
        idx_y = binary_search(ar, y + 1)
        # print(idx_x, idx_y)
        print(tree.get_sum(idx_y + 1, l - 1, r - 1) - tree.get_sum(idx_x + 1, l - 1, r - 1))

File: Lab_7/main2.cpp
//
// Created by rustam on 16.10.17.
//

#include <iostream>
#include <vector>

class SegmentTreeNode {
public:
    SegmentTreeNode(): val(0), left(NULL), right(NULL) {}
    SegmentTreeNode(int val, SegmentTreeNode* left, SegmentTreeNode* right) {
        this->val = val;
        this->left = left;
        this->right = right;
    }

    SegmentTreeNode operator+(const SegmentTreeNode& s) {
        return SegmentTreeNode(
                this->val + s.val,
                this,
                &s
        );
    }

    SegmentTreeNode& operator=(const SegmentTreeNode& s) {
        this->val = s.val;
        this->left = s.left;
        this->right = s.right;
        return *this;
    }
private:
    int val;
    SegmentTreeNode *left, *right;
};


class PersistentSegmentTree {
public:
    PersistentSegmentTree(const std::vector<uint32_t>& ar) {
        this->n = ar.size();
        this->tree = std::vector<std::vector<SegmentTreeNode>>(1, std::vector<SegmentTreeNode>(4 * this->n));
        this->_build_tree(ar, 1, 0, this->n - 1);
    }

    void set(uint32_t pos, uint32_t val) {
        this->tree.push_back(std::vector<uint32_t>());
        this->_set(1, 0, this->n - 1, pos, val);
    }

    uint32_t get_sum(uint32_t left, uint32_t right) {
        return this->_get_sum(1, 0, this->n - 1, left, right);
    }
private:
    SegmentTreeNode _build_tree_node(const std::vector<uint32_t>& ar, uint32_t idx) {
        uint32_t val = ar[idx];
        return SegmentTreeNode(
                val,
                NULL,
                NULL
        );
    }

    void _build_tree(const std::vector<uint32_t>& ar, uint32_t v, uint32_t tl, uint32_t tr) {
        if (tl == tr) {
            this->tree[v] = this->_build_tree_node(ar, tl);
        } else {
            int tm = (tl + tr) / 2;
            this->_build_tree(ar, 2 * v, tl, tm);
            this->_build_tree(ar, 2 * v + 1, tm + 1, tr);
            this->tree[0][v] = this->tree[0][2 * v] + this->tree[0][2 * v + 1];
        }
    }

    void _set(int v, int tl, int tr, int pos, long long val) {
        if (tl == tr) {
            this->tree.back().push_back(SegmentTreeNode(val, NULL, NULL));
        } else {
            int tm = (tr + tl) / 2;

            if (pos <= tm) {
                this->_set(2 * v, tl, tm, pos, val);


            } else {
                this->_set(2 * v + 1, tm + 1, tr, pos, val);
            }

            this->tree[v] = this->tree[2 * v] + this->tree[2 * v + 1];
        }
    }

    long long _get_sum(int v, int tl, int tr, int l, int r) {
        this->pushing(v);

        if (l > r) {
            return 0;
        }
        if ((l == tl) && (r == tr)) {
            return this->tree[v].sum_even;
        }

        int tm = (tr + tl) / 2;
        return this->_get_sum_even(2 * v, tl, tm, l, std::min(r, tm)) +
               this->_get_sum_even(2 * v + 1, tm + 1, tr, std::max(l, tm + 1), r);
    }

    int n;
    std::vector<std::vector<SegmentTreeNode>> tree;
};


int main() {
    int n, q;
    std::cin >> n >> q;


    std::vector<long long> A(n, 0);
    for (size_t i = 0; i < n; ++i) {
        std::cin >> A[i];
    }

    SegmentTree tree(A);

    int c, a, b;
    for (size_t i = 0; i < q; ++i) {
        std::cin >> c >> a >> b;
        map_operation(tree, c, a, b);

//        std::cout << c << " " << a << " " << b << "\n";
//        tree.print();
//        std::cout << "\n";
    }

    return 0;
}